### PR TITLE
Units prefix improvements

### PR DESCRIPTION
**Pull request by hsorby**
Looking at avoiding the exception raised by `std::stoi` this causes problems for web assembly binaries.  Hopefully other exceptions are harder to trigger.

Addresses #845.

### CODE REVIEW

File: src/units.cpp
Comment (nickerso) [inline src/units.cpp]: can't this use the new `convertPrefixToInt()` method?

Comment (hsorby) [reply src/units.cpp]: No because we don't know the state of the prefix in this method.  The `convertPrefixToInt()` method is for when the input is well formed.  That is, it is either a known prefix string or a valid CellML integer.

**Review by nickerso — DISMISSED** [review]
looks good once the clang formatting is kept happy...

**Review by nickerso — DISMISSED** [review]

File: src/units.cpp
Comment (agarny) [inline src/units.cpp]: I would use our new `convertPrefixToInt()` method for this, which means that `convertPrefixToInt()` would have to be reworked a bit to return `true` or `false` depending on whether the conversion was successful or not. IOW, I don't like "duplication" of code.

**Review by agarny — CHANGES_REQUESTED** [review]
Looks good besides some minor refactoring that I would like to see done.

File: src/units.cpp
Comment (hsorby) [reply src/units.cpp]: 9a0c7b9

File: src/units.cpp
Comment (agarny) [inline src/units.cpp]: Need a full stop.

File: src/units.cpp
Comment (agarny) [inline src/units.cpp]: I was expecting something like (not tested):
```c++
bool ok;

prefixMult = convertPrefixToInt(pre, &ok);

if (!ok) {
    return false;
}
```
with
```c++
int convertPrefixToInt(const std::string &in, bool *ok)
{
    int prefixInt = 0;

    if (ok != nullptr) {
        *ok = true;
    }

    if (isStandardPrefixName(in)) {
        prefixInt = standardPrefixList.at(in);
    } else if (in.empty()) {
        prefixInt = 0;
    } else if (isCellMLInteger(in)) {
        convertToInt(in, prefixInt);
    } else if (ok != nullptr) {
        *ok = false;
    }

    return prefixInt;
}
```

**Review by agarny — CHANGES_REQUESTED** [review]
I would go one extra step on that refactoring.

File: src/units.cpp
Comment (hsorby) [reply src/units.cpp]: My preference is to keep `convertPrefixToInt` as it is.  What do you think @nickerso?

File: src/units.cpp
Comment (hsorby) [reply src/units.cpp]: 90d15a2

File: src/units.cpp
Comment (agarny) [reply src/units.cpp]: You can still call `convertPrefixToInt()` the way you were doing it before, it's just that now you don't have to do those ugly tests before calling it in `updateUnitMultiplier()`. This makes code maintenance unnecessarily "complex".

Comment (hsorby) [reply src/units.cpp]: True, but I am also looking at the other locations that this function is used and that proposal doesn't work well for those cases.

Comment (nickerso) [reply src/units.cpp]: I've already approved this one twice, so I think its pretty clear I'm happy with it as is....

**Review by nickerso — DISMISSED** [review]

File: src/units.cpp
Comment (agarny) [reply src/units.cpp]: What other locations? Re `convertPrefixToInt()`, if you want the exact same behaviour as your version, then you could have:
```c++
int convertPrefixToInt(const std::string &in, bool *ok)
{
    int prefixInt = 0;

    if (ok != nullptr) {
        *ok = true;
    }

    if (isStandardPrefixName(in)) {
        prefixInt = standardPrefixList.at(in);
    } else if (in.empty()) {
        prefixInt = 0;
    } else if ((ok == nullptr) || isCellMLInteger(in)) {
        convertToInt(in, prefixInt);
    } else if (ok != nullptr) {
        *ok = false;
    }

    return prefixInt;
}
```

Comment (hsorby) [reply src/units.cpp]: `convertPrefixToInt` is used in the analyser and validator right.

Comment (agarny) [reply src/units.cpp]: Indeed, they use it and I have just tested my implementation at https://github.com/cellml/libcellml/pull/846#discussion_r618845525 and the one at https://github.com/cellml/libcellml/pull/846#discussion_r618896294, and it all works fine. So, yes, I would use my implementation at https://github.com/cellml/libcellml/pull/846#discussion_r618845525.

Comment (agarny) [reply src/units.cpp]: FWIW `convertPrefixToInt()` can be optimised a bit (since we initialise `prefixInt` to `0`:
```c++
int convertPrefixToInt(const std::string &in, bool *ok)
{
    int prefixInt = 0;

    if (ok != nullptr) {
        *ok = true;
    }

    if (isStandardPrefixName(in)) {
        prefixInt = standardPrefixList.at(in);
    } else if (isCellMLInteger(in)) {
        convertToInt(in, prefixInt);
    } else if ((ok != nullptr) && !in.empty()) {
        *ok = false;
    }

    return prefixInt;
}
```

Comment (hsorby) [reply src/units.cpp]: 7ebbfc9

**Review by agarny — DISMISSED** [review]

**Review by nickerso — APPROVED** [review]
Not a fan of the new `convertPrefixToInt()` method as it is now inconsistent with `convertToInt()` and `convertToReal()`. But otherwise this is good to go (assuming my changes to the documentation of that method are correct).

**Comment by agarny** [comment]
> Not a fan of the new `convertPrefixToInt()` method as it is now inconsistent with `convertToInt()` and `convertToReal()`.

Those are internal methods and the new `convertPrefixToInt()` method ensures easier maintenance of our code. If we really want, we could update `convertToInt()` and `convertToReal()` to also take an optional `ok` argument.

> But otherwise this is good to go (assuming my changes to the documentation of that method are correct).

Looks good to me, so approving this PR, but will leave it to @hsorby to merge it in after he has confirmed he's also happy with your change.

**Review by agarny — APPROVED** [review]

**Pull request merged**
